uniffi_core/src/ffi/rustfuture.rs

```diff
@@ -0,0 +1,642 @@
+//! [`Future::poll`]: https://doc.rust-lang.org/std/future/trait.Future.html#tymethod.poll
+//! [`Pol::Ready`]: https://doc.rust-lang.org/std/task/enum.Poll.html#variant.Ready
+//! [`Poll::Pending`]: https://doc.rust-lang.org/std/task/enum.Poll.html#variant.Pending
+//! [`Context`]: https://doc.rust-lang.org/std/task/struct.Context.html
+//! [`Waker`]: https://doc.rust-lang.org/std/task/struct.Waker.html
+//! [`RawWaker`]: https://doc.rust-lang.org/std/task/struct.RawWaker.html
+
+use super::FfiDefault;
+use crate::{call_with_result, FfiConverter, FfiError, FfiReturn, RustBuffer, RustCallStatus};
+use std::{
+    ffi::c_void,
+    future::Future,
+    mem::ManuallyDrop,
+    panic,
+    pin::Pin,
+    sync::Arc,
+    task::{Context, Poll, RawWaker, RawWakerVTable, Waker},
+};
+
+/// `RustFuture` represents a [`Future`] that can be sent over FFI safely-ish.
+///
+/// See the module documentation to learn more.
+#[repr(transparent)]
+pub struct RustFuture<T, E>(Pin<Box<dyn Future<Output = Result<T, E>> + 'static>>);
+
+impl<T, E> RustFuture<T, E> {
+    /// Create a new `RustFuture` from a regular `Future`.
+    pub fn new<F>(future: F) -> Self
+    where
+        F: Future<Output = Result<T, E>> + 'static,
+    {
+        Self(Box::pin(future))
+    }
+
+    /// Create a new `RustFuture` from a Tokio `Future`. It needs an
+    /// intermediate compatibility layer to be handlded as a regular Rust
+    /// `Future`.
+    #[cfg(feature = "tokio")]
+    pub fn new_tokio<F>(future: F) -> Self
+    where
+        F: Future<Output = Result<T, E>> + 'static,
+    {
+        Self(Box::pin(async_compat::Compat::new(future)))
+    }
+
+    /// Poll the future.
+    ///
+    /// There is one subtlety compared to `Future::poll` though: the
+    /// `foreign_waker` and `foreign_waker_environment` variables replace the
+    /// classical [`Context`]. We want the `RustFuture` **to be driven by the
+    /// foreign language's executor**. Hence the possibility for the foreign
+    /// language to provide its own waker function: Rust can signal something
+    /// has happened within the future and should be polled again.
+    ///
+    /// `Poll` is mapped to [`RustFuturePoll`].
+    ///
+    /// [`Context`]: https://doc.rust-lang.org/std/task/struct.Context.html
+    fn poll(
+        &mut self,
+        foreign_waker: RustFutureForeignWakerFunction,
+        foreign_waker_environment: *const RustFutureForeignWakerEnvironment,
+    ) -> RustFuturePoll<Result<T, E>> {
+        let waker = unsafe {
+            Waker::from_raw(RawWaker::new(
+                RustFutureForeignWaker::new(foreign_waker, foreign_waker_environment)
+                    .into_unit_ptr(),
+                &RustFutureForeignRawWaker::VTABLE,
+            ))
+        };
+        let mut context = Context::from_waker(&waker);
+
+        self.0.as_mut().poll(&mut context).into()
+    }
+}
+
+impl<T, E> FfiDefault for Option<Box<RustFuture<T, E>>> {
```

Comment:
This could probably be implemented more generically in `lib.rs`: `impl<T> FfiDefault for Option<T>`

Reply:
Yup!